### ISSUE TITLE
Allow HTTP::Client block initializer to be used when passing an URI

### ISSUE THREAD

**Issue by philnash** [crystal-lang/crystal#3492]
This doesn't work in `Crystal 0.19.4 (2016-10-21)`:

```crystal
require "http/client"

uri = URI.new("https://google.com")
HTTP::Client.new(uri) do |client|
  puts client.get("/").body
end
```

(Let me know if you want the full stack trace with it.)

But this does:

```crystal
require "http/client"

HTTP::Client.new("google.com", 443, true) do |client|
  puts client.get("/").body
end
```

I'd like to be able to use the block form when passing a URI as the argument to the initializer.

I'd also love to try to help implementing/fixing this. I don't have the time right now (I'm at work) but my guess is that I'd need to pass a dummy block argument to [this initializer](https://github.com/crystal-lang/crystal/blob/7f82f79bd1e3a7a1a73607e35c0662906736f1f8/src/http/client.cr#L151) and then [pass that through to the other initializer here](https://github.com/crystal-lang/crystal/blob/7f82f79bd1e3a7a1a73607e35c0662906736f1f8/src/http/client.cr#L154). I couldn't find anywhere else in the code (I took a look through `Enumerable`) where the dummy block argument was passed on though. If this is the right lines, let me know and I'll try to set up the project and fix it.

**Comment by philnash** [comment]
Sweet set of labels, thanks @asterite. I'm guessing I'm on the right path then. I've got the repo forked and the tests run so I'll see if I can have a go at this.

**Comment by luislavena** [comment]
This will be awesome to have @philnash!, as other have reported issues with `HTTP::Client.new` that was unexpected (recently #3455 comes to memory).

Looking forward your pull request! :smile:

**Comment by philnash** [comment]
So, this seemed to work but it doesn't feel right. https://github.com/philnash/crystal/commit/533eb8bd9dadb85a087bb39be03110acbd5d6f5b

I'm having trouble with the overloading, the types and the implicit blocks. Anyone got any hints?

**Comment by sdogruyol** [comment]
@philnash seems OK, why don't open a PR?

**Comment by philnash** [comment]
Thanks @sdogruyol, just opened #3521. Looking for feedback!

**Comment by luislavena** [comment]
Howdy, I believe this can be closed now that #3521 has been merged.

Cheers.

**Comment by philnash** [comment]
Agreed.
